README.md
Tpp One4All Library for PHP
=======================

[![Latest Version](https://img.shields.io/github/release/jilenloa/tppmyone4all-php.svg?style=flat-square)](https://github.com/jilenloa/tppmyone4all-php/releases)
[![Build Status](https://img.shields.io/travis/jilenloa/tppmyone4all-php.svg?style=flat-square)](https://travis-ci.org/jilenloa/tppmyone4all-php)
[![Total Downloads](https://img.shields.io/packagist/dt/jilenloa/tppmyone4all-php.svg?style=flat-square)](https://packagist.org/packages/jilenloa/tppmyone4all-php)

This library makes it easy for PHP developers to integrate with One4All TPP online service.

It's supported on PHP 5.5+

```php
use MyOne4All\TppClient;
use MyOne4All\Exceptions\TppException;
use MyOne4All\Models\DataBundle;
use MyOne4All\NetworkCodes;

$tppClient = new TppClient("apikey", "apisecret", "retailer");

echo $tppClient->getBalance(); # 0.0

// send airtime implementation
$airtime_response = $tppClient->sendAirtime("0245667942", 1, "trans03423423", NetworkCodes::AUTO_DETECT);
if($airtime_response->isSuccessful()){
    echo "airtime sent";
}else{
    echo "Failed: ".$airtime_response->getErrorMessage();
}

// send internet data implementation
$data_code = "DAILY_20MB";
$transaction_reference = "trans03423423";
$bundle_response = $tppClient->sendDataBundle("0245667942", $data_code, $transaction_reference, NetworkCodes::MTN_GH);
if($bundle_response->isSuccessful()){
    echo "internet bundle sent";
}else{
    echo "Failed: ".$bundle_response->getErrorMessage();
}


// send mobile money implementation
$transaction_reference = "trans03423423";
$amount = 1;
$momo_response = $tppClient->sendMobileMoney("0245667942", $amount, $transaction_reference);
if($momo_response->isSuccessful()){
    echo "mobile money sent";
}else if($momo_response->isPending()){
    // this is very important, do not treat as an error.
    // you must check transaction status later to know if it was successful or not
    echo "mobile money request is being processed.";
}else{
    echo "Failed: ".$momo_response->getErrorMessage();
}

// receive mobile money implementation
$transaction_reference = "trans03423423";
$amount = 1;
$payer_number = "0245667XXX";
$momo_response = $tppClient->receiveMobileMoney($payer_number, $amount, $transaction_reference);
if($momo_response->isSuccessful()){
    // check transaction status later to confirm receipt
    echo "mobile money payment request initiated";
}else{
    echo "Failed: ".$momo_response->getErrorMessage();
}

$sms_message = "hello world";
$sms_sender_id = "One4All";
$transaction_reference = "sms11";
$sms_response = $tppClient->sendSms("0572180376", $sms_message, $sms_sender_id, $transaction_reference);
if($sms_response->isSuccessful()){
    echo "sms message queued";
}else{
    echo "Failed: ".$sms_response->getErrorMessage();
}



echo NetworkCodes::AUTO_DETECT; # 0
echo NetworkCodes::MTN_GH; # 4
echo NetworkCodes::AIRTEL_DRC; # 2

// fetch available data packages for all networks
$data_packages = $tppClient->getDataBundleList();

// only fetch data packages for mtn
$data_packages = $tppClient->getDataBundleList(NetworkCodes::MTN_GH);

foreach($data_pages as $data_package){
    echo $data_package->plan_id; # DAILY_20MB
    echo $data_package->plan_name; # DAILY_20MB
    echo $data_package->category; # DAILY
    echo $data_package->network_id; # 4
    echo $data_package->volume; # 20 MB
}
```

Install package using Composer
```bash
composer require jilenloa/tppmyone4all-php
```

After installing, you need to require Composer's autoloader:

```php
require 'vendor/autoload.php';
```

You can then later update the package later using composer:

 ```bash
composer update
 ```

In order to run unit tests for this package

 ```bash
npm install -g mockserver
 ```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.5](https://github.com/jilenloa/tppmyone4all-php/compare/v0.0.4...v0.0.5) (2019-10-28)

### [0.0.4](https://github.com/jilenloa/tppmyone4all-php/compare/v0.0.3...v0.0.4) (2019-10-28)

### [0.0.3](https://github.com/jilenloa/tppmyone4all-php/compare/v0.0.2...v0.0.3) (2019-10-28)

### [0.0.2](https://github.com/jilenloa/tppmyone4all-php/compare/v0.0.1...v0.0.2) (2019-10-28)

### 0.0.1 (2019-10-28)


tests/Server.php
<?php
namespace MyOne4All\Tests;
use GuzzleHttp\Client;
use GuzzleHttp\Psr7;
use Psr\Http\Message\ResponseInterface;
use Symfony\Component\Process\Process;

/**
 * The Server class is used to control a scripted webserver using node.js that
 * will respond to HTTP requests with queued responses.
 *
 * Queued responses will be served to requests using a FIFO order.  All requests
 * received by the server are stored on the node.js server and can be retrieved
 * by calling {@see Server::received()}.
 *
 * Mock responses that don't require data to be transmitted over HTTP a great
 * for testing.  Mock response, however, cannot test the actual sending of an
 * HTTP request using cURL.  This test server allows the simulation of any
 * number of HTTP request response transactions to test the actual sending of
 * requests over the wire without having to leave an internal network.
 */
class Server
{
    /** @var Client */
    private static $client;
    private static $started = false;
    public static $url = 'http://127.0.0.1:9005/'; //mockserver -p 9005 -m mocks
    public static $port = 9005;

    public static function wait($maxTries = 5)
    {
        $tries = 0;
        while (!self::isListening() && ++$tries < $maxTries) {
            usleep(100000);
        }
        if (!self::isListening()) {
            throw new \RuntimeException('Unable to contact node.js mockserver server on port '.self::$port);
        }
    }
    public static function start()
    {
        if (self::$started) {
            return;
        }
        if (!self::isListening()) {

            $command = 'mockserver -p 9005 -m ' . __DIR__ . '/mocks';
            $process = new Process($command);
            $process->start();

            self::wait();
        }
        self::$started = true;
    }
    private static function isListening()
    {
        try {
            self::getClient()->request('GET', 'server', [
                'connect_timeout' => 5,
                'timeout'         => 5
            ]);
            return true;
        } catch (\Exception $e) {
            return false;
        }
    }
    private static function getClient()
    {
        if (!self::$client) {
            self::$client = new Client([
                'base_uri' => self::$url,
                'sync'     => true,
            ]);
        }
        return self::$client;
    }
}

tests/bootstrap.php
<?php

namespace MyOne4All\Test {
    require __DIR__ . '/../vendor/autoload.php';
    require __DIR__ . '/Server.php';
    use MyOne4All\Tests\Server;
    Server::start();

    //register_shutdown_function(function () {
        //Server::stop();
    //});
}

tests/TppClientTest.php
<?php
/**
 * Created by PhpStorm.
 * User: macbookpro
 * Date: 10/27/19
 * Time: 9:47 PM
 */

namespace MyOne4All\Tests;


use MyOne4All\TppClient;
use MyOne4All\Exceptions\TppException;
use MyOne4All\Models\DataBundle;
use MyOne4All\NetworkCodes;
use MyOne4All\TppResponse;
use PHPUnit\Framework\TestCase;

class TppClientTest extends TestCase
{
    /** @var TppClient */
    protected $tppClient;

    protected function setUp()
    {
        parent::setUp();
        $this->tppClient = new TppClient("testkey", "testsecret", "testretailer", TppClient::COUNTRY_GHANA);
        $this->tppClient->enableMock();
    }

    public function testAirtime(){
        $response = $this->tppClient->sendAirtime("0245667942", 1, "809003324634425623", NetworkCodes::AUTO_DETECT);
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful());
    }


    public function testAirtimeFail(){
        $response = $this->tppClient->sendAirtime("02456679", 1, "809003324634425623", NetworkCodes::AUTO_DETECT);
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->failed());
        $this->assertFalse($response->isSuccessful());
        $this->assertNotEmpty($response->getErrorMessage());
        $this->assertTrue(is_string($response->getErrorMessage()));
    }

    public function testBalance(){
        $response = $this->tppClient->getBalance();
        $this->assertNotNull($response);
        $this->assertTrue(is_numeric($response));
    }

    public function testSendDataBundle(){
        $response = $this->tppClient->sendDataBundle("0245667942", "DAILY_20MB", "809003324634425624", NetworkCodes::MTN_GH);
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful());
    }

    public function testQueryBundle(){
        $response = $this->tppClient->getDataBundleList();
        $this->assertTrue(is_array($response));
        $this->assertNotNull(current($response));
        $this->assertInstanceOf(DataBundle::class, current($response));
    }

    /**
     * @throws TppException
     */
    public function testTransactionStatusCheck(){
        $response = $this->tppClient->getTransactionStatus("809003324634425623");
        $this->assertTrue(is_string($response));
        $this->assertNotEmpty($response);
    }

    /**
     * @throws TppException
     */
    public function testTransactionNotFound(){
        $this->setExpectedException(TppException::class, 'Transaction not Found');
        $this->tppClient->getTransactionStatus("8090034634425623");
    }

    public function testC2b(){
        $response = $this->tppClient->receiveMobileMoney("0245667942", 1, "809003324634425623");
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful());
    }

    public function testB2cMTNGhana(){
        $response = $this->tppClient->sendMobileMoney("0245667942", 1, "809003324634425623");
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful());
        $this->assertFalse($response->isPending());
    }

    public function testB2cAirtelTigoGhana(){
        $response = $this->tppClient->sendMobileMoney("0572180376", 1, "809003324634425623");
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful() || $response->isPending());
        $this->assertFalse($response->isSuccessful());
        $this->assertTrue($response->isPending());
        $this->assertFalse($response->failed());
        $this->assertNull($response->getErrorMessage());
    }

    public function testB2cFail(){
        $response = $this->tppClient->sendMobileMoney("057218037", 1, "809003324634425623");
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertFalse($response->isSuccessful() || $response->isPending());
        $this->assertFalse($response->isSuccessful());
        $this->assertFalse($response->isPending());
        $this->assertNotNull($response->getErrorMessage());
        $this->assertNotEmpty($response->getErrorMessage());
        $this->assertTrue($response->failed());
    }

    public function testSms(){
        $response = $this->tppClient->sendSms("0572180376", "hello world", "One4All", "sms11");
        $this->assertInstanceOf(TppResponse::class, $response);
        $this->assertTrue($response->isSuccessful());
    }

}

src/TppResponse.php
<?php
/**
 * Created by PhpStorm.
 * User: macbookpro
 * Date: 10/27/19
 * Time: 3:40 PM
 */

namespace MyOne4All;


class TppResponse implements \JsonSerializable
{

    protected $response_array;
    /**
     * TppResponse constructor.
     * @param mixed $response_array
     */
    public function __construct($response_array)
    {
        $this->response_array = $response_array;
    }

    /**
     * Specify data which should be serialized to JSON
     * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
     * @return mixed data which can be serialized by <b>json_encode</b>,
     * which is a value of any type other than a resource.
     * @since 5.4.0
     */
    public function jsonSerialize()
    {
        return $this->response_array;
    }

    public function isSuccessful(){
        if($this->response_array['status-code'] === "00"){
            if(isset($this->response_array['pending'])){
                return !$this->response_array['pending'];
            }
            return true;
        }else{
            return false;
        }
    }

    public function isPending(){
        if(isset($this->response_array['pending'])){
            return $this->response_array['pending'];
        }
        return false;
    }

    public function getErrorMessage(){
        if($this->failed()){
            return $this->response_array['message'];
        }
        return null;
    }

    public function getMessage(){
        return $this->response_array['message'];
    }

    public function failed(){
        return !$this->isSuccessful() && !$this->isPending();
    }

    public function getBalanceBefore(){
        if(isset($this->response_array['balance_before'])){
            return $this->response_array['balance_before'];
        }
        return null;
    }

    public function getBalanceAfter(){
        if(isset($this->response_array['balance_after'])){
            return $this->response_array['balance_after'];
        }
        return null;
    }
}

src/TppClient.php
<?php
/**
 * Created by PhpStorm.
 * User: macbookpro
 * Date: 10/27/19
 * Time: 3:00 PM
 */

namespace MyOne4All;


use GuzzleHttp\Exception\ClientException;
use MyOne4All\Exceptions\TppException;
use MyOne4All\Models\DataBundle;

class TppClient
{
    private $key;
    private $secret;
    private $retailer;
    private $country;
    protected $httpClient;

    private $mock;

    const COUNTRY_GHANA = 'gh';
    const COUNTRY_CONGO = 'drc';

    /**
     * Client constructor.
     * @param $key
     * @param $secret
     * @param $retailer
     * @param string $country
     */
    public function __construct($key, $secret, $retailer, $country = self::COUNTRY_GHANA)
    {
        $this->key = $key;
        $this->secret = $secret;
        $this->retailer = $retailer;
        $this->country = $country;
    }

    private function getHttpClient(){
        if($this->httpClient){
            return $this->httpClient;
        }

        $base_uri = "https://tpp{$this->country}.myone4all.com";
        if($this->mock){
            $base_uri = "http://127.0.0.1:9005";
        }

        $this->httpClient = new \GuzzleHttp\Client(
            array(
                'base_uri' => $base_uri,
                'headers' => array(
                    'ApiKey' => $this->key,
                    'ApiSecret' => $this->secret,
                    'User-Agent' => 'TppMyOne4All PHP Library',
                    'Accept' => 'application/json'
                ),
                'verify' => false
            )
        );

        return $this->httpClient;
    }

    /**
     * @return string|double
     * @throws ClientException
     */
    public function getBalance(){
        $response = $this->getHttpClient()->get("/api/TopUpApi/balance");
        $response_array = json_decode($response->getBody()->getContents(), true);
        return $response_array['balance'];
    }

    /**
     * @param $recipient
     * @param $amount
     * @param $transaction_reference
     * @param int $network
     * @return TppResponse
     * @throws ClientException
     */
    public function sendAirtime($recipient, $amount, $transaction_reference, $network = NetworkCodes::AUTO_DETECT){
        $params = array(
            'retailer' => $this->retailer,
            'recipient' => $recipient,
            'amount' => $amount,
            'network' => $network,
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/airtime?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);
        return new TppResponse($response_array);
    }

    /**
     * @param $recipient
     * @param $data_code
     * @param $transaction_reference
     * @param $network
     * @return TppResponse
     * @throws ClientException
     */
    public function sendDataBundle($recipient, $data_code, $transaction_reference, $network){
        $params = array(
            'retailer' => $this->retailer,
            'recipient' => $recipient,
            'data_code' => $data_code,
            'network' => $network,
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/dataBundle?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);
        return new TppResponse($response_array);
    }

    /**
     * @param $recipient
     * @param $amount
     * @param $transaction_reference
     * @return TppResponse
     * @throws ClientException
     */
    public function sendMobileMoney($recipient, $amount, $transaction_reference){
        $params = array(
            'amount' => $amount,
            'recipient' => $recipient,
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/b2c?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);
        return new TppResponse($response_array);
    }

    /**
     * @param $payer_number
     * @param $amount
     * @param $transaction_reference
     * @return TppResponse
     * @throws ClientException
     */
    public function receiveMobileMoney($payer_number, $amount, $transaction_reference){
        $params = array(
            'amount' => $amount,
            'recipient' => $payer_number,
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/c2b?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);
        return new TppResponse($response_array);
    }

    /**
     * @param $transaction_reference
     * @return string
     * @throws ClientException
     * @throws TppException
     */
    public function getTransactionStatus($transaction_reference){
        $params = array(
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/transactionStatus?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);

        if(isset($response_array['transaction-state'])){
            return $response_array['transaction-state'];
        }else{
            if(isset($response_array['message'])){
                throw new TppException($response_array['message']);
            }else{
                throw new TppException("Unknown response from Tpp server");
            }
        }

    }

    /**
     * @param null $network
     * @return array|DataBundle[]
     * @throws ClientException
     */
    public function getDataBundleList($network = null){
        if($network){
            $params = array(
                'network' => $network
            );
        }else{
            $params = array();
        }

        $response = $this->getHttpClient()->get("/api/TopUpApi/dataBundleList?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);

        $list = array();
        foreach ($response_array['bundles'] as $item) {
            $list[] = new DataBundle($item);
        }

        return $list;
    }

    public function sendSms($recipient, $message, $sender_id, $transaction_reference){
        $params = array(
            'message' => $message,
            'recipient' => $recipient,
            'sender_id' => $sender_id,
            'trxn' => $transaction_reference
        );
        $response = $this->getHttpClient()->get("/api/TopUpApi/sms?".http_build_query($params));
        $response_array = json_decode($response->getBody()->getContents(), true);
        return new TppResponse($response_array);
    }

    public function enableMock()
    {
        $this->mock = true;
    }


}